In [4]:
import random

import altair as alt
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.compose import make_column_transformer
from sklearn.compose import make_column_selector
from sklearn.neighbors import KNeighborsClassifier
from sklearn.pipeline import make_pipeline
from sklearn.model_selection import train_test_split
from sklearn.metrics.pairwise import euclidean_distances
from sklearn import set_config
from sklearn.metrics.pairwise import euclidean_distances
from sklearn.model_selection import (
    GridSearchCV,
    RandomizedSearchCV,
    cross_validate,
    train_test_split,
)

# Simplify working with large datasets in Altair
alt.data_transformers.disable_max_rows()

# Output dataframes instead of arrays
set_config(transform_output="pandas")

In [5]:
pip install ucimlrepo

Note: you may need to restart the kernel to use updated packages.


In [6]:
from ucimlrepo import fetch_ucirepo 
# fetch dataset 
htru2 = fetch_ucirepo(id=372) 
  
# data (as pandas dataframes) 
X = htru2.data.features 
y = htru2.data.targets 
  
# metadata 
print(htru2.metadata) 
  
# variable information 
print(htru2.variables) 

{'uci_id': 372, 'name': 'HTRU2', 'repository_url': 'https://archive.ics.uci.edu/dataset/372/htru2', 'data_url': 'https://archive.ics.uci.edu/static/public/372/data.csv', 'abstract': 'Pulsar candidates collected during the HTRU survey. Pulsars are a type of star, of considerable scientific interest. Candidates must be classified in to pulsar and non-pulsar classes to aid discovery.', 'area': 'Physics and Chemistry', 'tasks': ['Classification', 'Clustering'], 'characteristics': ['Multivariate'], 'num_instances': 17898, 'num_features': 8, 'feature_types': ['Real'], 'demographics': [], 'target_col': ['class'], 'index_col': None, 'has_missing_values': 'no', 'missing_values_symbol': None, 'year_of_dataset_creation': 2015, 'last_updated': 'Wed Apr 03 2024', 'dataset_doi': '10.24432/C5DK6R', 'creators': ['Robert Lyon'], 'intro_paper': {'title': 'Fifty years of pulsar candidate selection: from simple filters to a new principled real-time classification approach', 'authors': 'R. Lyon, B. Stapper

In [7]:
star_data = pd.merge(X, y, left_index=True,right_index=True)
star_data

Profile_mean  Profile_stdev  Profile_skewness  Profile_kurtosis  \
0        140.562500      55.683782         -0.234571         -0.699648   
1        102.507812      58.882430          0.465318         -0.515088   
2        103.015625      39.341649          0.323328          1.051164   
3        136.750000      57.178449         -0.068415         -0.636238   
4         88.726562      40.672225          0.600866          1.123492   
...             ...            ...               ...               ...   
17893    136.429688      59.847421         -0.187846         -0.738123   
17894    122.554688      49.485605          0.127978          0.323061   
17895    119.335938      59.935939          0.159363         -0.743025   
17896    114.507812      53.902400          0.201161         -0.024789   
17897     57.062500      85.797340          1.406391          0.089520   

          DM_mean   DM_stdev  DM_skewness  DM_kurtosis  class  
0        3.199833  19.110426     7.975532    74.242225      0  
1        1.677258  14.860146    10.576487   127.393580      0  
2        3.121237  21.744669     7.735822    63.171909      0  
3        3.642977  20.959280     6.896499    53.593661      0  
4        1.178930  11.468720    14.269573   252.567306      0  
...           ...        ...          ...          ...    ...  
17893    1.296823  12.166062    15.450260   285.931022      0  
17894   16.409699  44.626893     2.945244     8.297092      0  
17895   21.430602  58.872000     2.499517     4.595173      0  
17896    1.946488  13.381731    10.007967   134.238910      0  
17897  188.306020  64.712562    -1.597527     1.429475      0  

[17898 rows x 9 columns]

In [8]:
np.random.seed(2024)

star_train, star_test = train_test_split(
    star_data, train_size=0.80,
)
star_train

Profile_mean  Profile_stdev  Profile_skewness  Profile_kurtosis  \
14117     89.726562      42.207026          0.414897          0.904615   
3787     116.960938      73.153990          0.316964         -1.166485   
765      136.664062      52.628307          0.067857         -0.492523   
9684     116.046875      43.673837          0.330900          0.308270   
8149     118.843750      48.989865          0.335433          0.134276   
...             ...            ...               ...               ...   
16567    128.125000      48.357555         -0.005863         -0.100869   
2494     135.554688      41.715706         -0.047587          0.910534   
14875    124.976562      45.683946         -0.209657          0.109942   
2688     125.304688      49.947873          0.053109         -0.068939   
7816     120.867188      39.129267          0.068661          1.005908   

          DM_mean   DM_stdev  DM_skewness  DM_kurtosis  class  
14117    2.403846  13.488690     9.129472   121.734406      0  
3787   201.534281  47.262251    -1.704449     2.788871      0  
765      4.382107  24.805369     6.734455    49.054326      0  
9684     3.107023  20.456870     7.857742    69.309930      0  
8149     0.775084  11.946176    17.115458   312.061142      0  
...           ...        ...          ...          ...    ...  
16567    2.019231  15.440807    10.652478   137.918490      0  
2494     2.617893  17.409786     8.831467    89.780556      0  
14875    4.981605  26.593647     5.563959    31.319089      0  
2688     1.628763  12.247147    12.262394   195.921439      0  
7816     2.090301  13.255599    10.760405   152.782160      0  

[14318 rows x 9 columns]

In [9]:
star_preprocessor = make_column_transformer(
    (StandardScaler(), ["Profile_mean", "Profile_stdev", "Profile_skewness"
                        ,"Profile_kurtosis", "DM_mean", "DM_stdev", "DM_skewness"
                        ,"DM_kurtosis"]),
     verbose_feature_names_out=False
)
star_preprocessor

ColumnTransformer(transformers=[('standardscaler', StandardScaler(),
                                 ['Profile_mean', 'Profile_stdev',
                                  'Profile_skewness', 'Profile_kurtosis',
                                  'DM_mean', 'DM_stdev', 'DM_skewness',
                                  'DM_kurtosis'])],
                  verbose_feature_names_out=False)

In [10]:
param_grid = {
    "kneighborsclassifier__n_neighbors": range(2, 15, 1),
}
star_pipe = make_pipeline(star_preprocessor, KNeighborsClassifier())

In [11]:
knn_star_grid = GridSearchCV(
    estimator=star_pipe, 
    param_grid=param_grid, 
    cv=5,
    n_jobs=-1,
    return_train_score=True
)
knn_star_grid

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('columntransformer',
                                        ColumnTransformer(transformers=[('standardscaler',
                                                                         StandardScaler(),
                                                                         ['Profile_mean',
                                                                          'Profile_stdev',
                                                                          'Profile_skewness',
                                                                          'Profile_kurtosis',
                                                                          'DM_mean',
                                                                          'DM_stdev',
                                                                          'DM_skewness',
                                                                          'DM_kurtosis'])],
                                                          verbose_feature_names_out=False)),
                                       ('kneighborsclassifier',
                                        KNeighborsClassifier())]),
             n_jobs=-1,
             param_grid={'kneighborsclassifier__n_neighbors': range(2, 15)},
             return_train_score=True)

In [12]:
X = star_train.drop(columns=["class"])
y = star_train["class"]

knn_model_grid = knn_star_grid.fit(X, y)

accuracies_grid = pd.DataFrame(knn_model_grid.cv_results_)
accuracies_grid

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0        0.013726      0.001114         0.137483        0.018047   
1        0.013032      0.000136         0.134177        0.005996   
2        0.013081      0.000160         0.134785        0.003539   
3        0.013069      0.000090         0.138191        0.004928   
4        0.013098      0.000105         0.147168        0.015533   
5        0.013181      0.000138         0.140737        0.004075   
6        0.013557      0.000802         0.147095        0.014772   
7        0.013177      0.000069         0.142993        0.003876   
8        0.013286      0.000107         0.145514        0.004434   
9        0.013303      0.000124         0.146282        0.004254   
10       0.013329      0.000106         0.165261        0.031991   
11       0.019899      0.013134         0.158828        0.019051   
12       0.014310      0.000544         0.150108        0.004660   

   param_kneighborsclassifier__n_neighbors  \
0                                        2   
1                                        3   
2                                        4   
3                                        5   
4                                        6   
5                                        7   
6                                        8   
7                                        9   
8                                       10   
9                                       11   
10                                      12   
11                                      13   
12                                      14   

                                       params  split0_test_score  \
0    {'kneighborsclassifier__n_neighbors': 2}           0.975908   
1    {'kneighborsclassifier__n_neighbors': 3}           0.979050   
2    {'kneighborsclassifier__n_neighbors': 4}           0.978003   
3    {'kneighborsclassifier__n_neighbors': 5}           0.979050   
4    {'kneighborsclassifier__n_neighbors': 6}           0.978701   
5    {'kneighborsclassifier__n_neighbors': 7}           0.980098   
6    {'kneighborsclassifier__n_neighbors': 8}           0.978352   
7    {'kneighborsclassifier__n_neighbors': 9}           0.978003   
8   {'kneighborsclassifier__n_neighbors': 10}           0.978352   
9   {'kneighborsclassifier__n_neighbors': 11}           0.978352   
10  {'kneighborsclassifier__n_neighbors': 12}           0.978701   
11  {'kneighborsclassifier__n_neighbors': 13}           0.979050   
12  {'kneighborsclassifier__n_neighbors': 14}           0.978352   

    split1_test_score  split2_test_score  split3_test_score  ...  \
0            0.974511           0.978352           0.975899  ...   
1            0.976257           0.979749           0.977995  ...   
2            0.976257           0.976955           0.977297  ...   
3            0.976257           0.979050           0.977646  ...   
4            0.974860           0.977654           0.977297  ...   
5            0.975559           0.978352           0.977646  ...   
6            0.974860           0.977654           0.978694  ...   
7            0.976606           0.979050           0.978344  ...   
8            0.975908           0.979050           0.977995  ...   
9            0.976257           0.979050           0.978344  ...   
10           0.975559           0.979749           0.977995  ...   
11           0.976257           0.979749           0.978344  ...   
12           0.975209           0.979399           0.977995  ...   

    mean_test_score  std_test_score  rank_test_score  split0_train_score  \
0          0.975346        0.002058               13            0.984983   
1          0.977441        0.002022                8            0.983325   
2          0.976812        0.000845               12            0.980967   
3          0.977720        0.001178                3            0.981404   
4          0.977162        0.001261               10            0.980007   
5          0.977511        0.001663         

In [25]:
accuracy_versus_k_grid = alt.Chart(accuracies_grid, title = "accuracy of different K value" ).mark_line(point=True).encode(
     x=alt.X("param_kneighborsclassifier__n_neighbors")
         .title("Neighbors")
         .scale(zero=False),
     y=alt.Y("mean_test_score")
         .title("Accuracy estimate")
         .scale(zero=False)
 ).properties(
    width=800
)

accuracy_versus_k_grid

alt.Chart(...)

From the plot above, we can see that K = 13 provides the highest accuracy.
So that we could use K = 13 to estimate the class.

In [16]:
knn_spec = KNeighborsClassifier(n_neighbors=13)
star_pipe_final = make_pipeline(star_preprocessor,knn_spec)

X = star_train.drop(columns=["class"])
y = star_train["class"]

star_final_fit = star_pipe_final.fit(X,y)
star_final_fit

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(transformers=[('standardscaler',
                                                  StandardScaler(),
                                                  ['Profile_mean',
                                                   'Profile_stdev',
                                                   'Profile_skewness',
                                                   'Profile_kurtosis',
                                                   'DM_mean', 'DM_stdev',
                                                   'DM_skewness',
                                                   'DM_kurtosis'])],
                                   verbose_feature_names_out=False)),
                ('kneighborsclassifier', KNeighborsClassifier(n_neighbors=13))])

In [19]:
star_test_predictions = star_test.assign(
    predicted=star_final_fit.predict(star_test[["Profile_mean", "Profile_stdev", 
                                     "Profile_skewness","Profile_kurtosis", "DM_mean", 
                                     "DM_stdev", "DM_skewness" ,"DM_kurtosis"]])
)
star_test_predictions

Profile_mean  Profile_stdev  Profile_skewness  Profile_kurtosis  \
12886     88.148438      39.451729          0.605601          1.624632   
8692     109.453125      48.281055          0.126580         -0.250872   
17650    144.242188      49.522840         -0.108100          0.165935   
1781     124.500000      53.521608          0.034350         -0.245293   
13889    100.312500      48.392398          0.714821          0.683510   
...             ...            ...               ...               ...   
10090    136.125000      57.965237          0.157111         -0.555772   
8559     116.546875      49.238079          0.388811         -0.108516   
5129     131.671875      51.689444         -0.083764         -0.377586   
15693    107.539062      60.053479          0.049761         -0.806445   
6006     109.164062      46.074291          0.258365          0.657396   

         DM_mean   DM_stdev  DM_skewness  DM_kurtosis  class  predicted  
12886   2.527592  15.288149     9.307637   110.292943      0          0  
8692    2.988294  20.520604     7.790484    64.666993      0          0  
17650  11.497492  45.517766     4.036984    15.098706      0          0  
1781    2.151338  17.634329     9.624466   101.933256      0          0  
13889   2.263378  18.163970     9.967299   107.651216      0          0  
...          ...        ...          ...          ...    ...        ...  
10090   2.382107  16.117731     9.162124   100.398098      0          0  
8559    0.999164  11.771195    16.214952   299.638780      0          0  
5129    7.989967  34.976269     4.544575    20.265613      0          0  
15693   1.454849  11.566102    12.762448   220.034326      0          0  
6006    1.003344  11.414580    15.802571   293.475994      0          0  

[3580 rows x 10 columns]

In [20]:
X_test = star_test[["Profile_mean", "Profile_stdev", 
                                     "Profile_skewness","Profile_kurtosis", "DM_mean", 
                                     "DM_stdev", "DM_skewness" ,"DM_kurtosis"]]
y_test = star_test["class"]

star_prediction_accuracy = star_final_fit.score(X_test, y_test)
star_prediction_accuracy

0.9815642458100559

The accuracy is around 98%.

In [21]:
star_mat = pd.crosstab(
    star_test_predictions["class"],  # True labels
    star_test_predictions["predicted"],  # Predicted labels
)
star_mat

predicted     0    1
class               
0          3232   20
1            46  282

In [ ]:
Then, we assign one label as positive and another as negative.
Here we are more interested in 1 so we assign it as positive.

In [22]:
precision = 282 / ( 282 + 20)
precision

0.9337748344370861

In [23]:
recall = 282 / (282 + 46)
recall

0.8597560975609756

The precision is about 93% and the recall is about 86%.